unicorn/src/cluster/unicorn.cpp

```diff
@@ -0,0 +1,285 @@
+    auto it = storage->find(uuid);
+    bool drop = false;
+    if(it != storage->end()) {
+        storage->erase(it);
+        drop = true;
+    }
+    if(drop) {
+        parent->locator.drop_node(uuid);
+    }
+}
+
+unicorn_cluster_t::on_list_update::on_list_update(unicorn_cluster_t* _parent) :
+parent(_parent)
+{}
+
+void
+unicorn_cluster_t::on_list_update::write(unicorn::api_t::response::children_subscribe_result&& result) {
+    auto nodes = std::get<1>(result);
+    std::sort(nodes.begin(), nodes.end());
+    std::vector<std::string> to_delete, to_add;
+    auto storage = parent->registered_locators.synchronize();
+    std::set_difference(nodes.begin(), nodes.end(), storage->begin(), storage->end(), std::back_inserter(to_add));
+    std::set_difference(storage->begin(), storage->end(), nodes.begin(), nodes.end(), std::back_inserter(to_delete));
+    for(size_t i = 0; i < to_delete.size(); i++) {
+        parent->locator.drop_node(to_delete[i]);
+        storage->erase(to_delete[i]);
+    }
+    for(size_t i = 0; i < to_add.size(); i++) {
+        if(to_add[i] == parent->locator.uuid()) {
+            continue;
+        }
+        parent->unicorn.get(
+            std::make_shared<on_fetch>(to_add[i], parent),
+            parent->config.path + '/' + to_add[i]
+        );
+    }
+}
+
+void
+unicorn_cluster_t::on_list_update::abort(int rc, const std::string& reason) {
+    COCAINE_LOG_WARNING(parent->log, "FAILURE during subscription(%i): %s. Resubscribing.", rc, reason.c_str());
+    parent->subscribe_timer.expires_from_now(boost::posix_time::seconds(parent->config.retry_interval));
+    parent->subscribe_timer.async_wait(std::bind(&unicorn_cluster_t::on_subscribe_timer, parent, std::placeholders::_1));
+
+}
+
+unicorn_cluster_t::unicorn_cluster_t(
+    cocaine::context_t & _context,
+    cocaine::api::cluster_t::interface & _locator,
+    const std::string& name,
+    const cocaine::dynamic_t& args
+):
+    cluster_t(_context, _locator, name, args),
+    log(_context.log("unicorn")),
+    config(args),
+    context(_context),
+    locator(_locator),
+    announce_timer(_locator.asio()),
+    subscribe_timer(_locator.asio()),
+    zk_session(),
+    zk(unicorn::make_zk_config(args), zk_session),
+    unicorn(*log, zk)
+{
+    subscribe_timer.expires_from_now(boost::posix_time::seconds(config.retry_interval));
+    subscribe_timer.async_wait(std::bind(&unicorn_cluster_t::on_subscribe_timer, this, std::placeholders::_1));
+
+    announce_timer.expires_from_now(boost::posix_time::seconds(config.retry_interval));
+    announce_timer.async_wait(std::bind(&unicorn_cluster_t::on_announce_timer, this, std::placeholders::_1));
+}
+
+void
+unicorn_cluster_t::announce() {
+    const auto& actor = context.locate("locator");
+
+    if(!actor) {
+        COCAINE_LOG_ERROR(log, "unable to announce local endpoints: locator is not available");
+        announce_timer.expires_from_now(boost::posix_time::seconds(config.retry_interval));
```

Comment:
If it's not there, it won't appear later. Actually, how about we use the name passed to you instead of hardcoded "locator"? Right now it's something like "locator:cluster" but I guess we could change it around so that the name passed to the constructor is the actual name of the locator service.